File: pkg/R/closest.r
## Given an input grid of x,y,z pairs
## and a set of new x,y, pairs
## get the z of closest location on grid to the input x,y

## LLNL-CODE-637312

## (see also aterp.r)

## 8/25/10
## rename the arguments so that names make it obvious  (did this earlier for aterp.r)
## which is in input (interpolate from) and which is the
## output (interpolate to)
##    xy.out     was   xy0
##    xyz.inp    was   iout  

## arguments are
##  xy.out   list(x=,y=)
##        points for which we need to interpolate a z value
##
##  xyz.inp    list with a structure suitable for use in image()
##          vectors x,y that define a grid
##          matrix z of values on that grid
##          e.g.:
##            list returned by cap.add()
##            list returned by interp()


## fortran subroutine assumes that all points in xy.out are inside the
## interpolated region
## if xy.out$x and xy.out$y are not the same length, use the shorter

## use f77 -c -cg89 closest.f

closest <- function(xyz.inp,xy.out,info=FALSE) {

  ## 4/25/08 NOTE: fix for points in xy.out outside the range where the grid has data
  ## has not been tested
  cat('[closest.r] WARNING:  not tested for xy.out points outside range of data\n')
  
  ## have to make the location of the fortran code more general
  ##  if (!is.loaded(symbol.For("closest")))
  ##  print(dyn.load( paste(getenv('SPLUSLIB'),'/delaunay/deldir.o',sep='') ))
  ##    print(dyn.load('/us/macq/epd/neshaps/afuns/closest.o'))
  
  ## later, make a package, then this will be automatic
  ##  if (!is.loaded('closest')) dyn.load('/home/macqueen1/epd/neshaps/afuns/closest.so')
  
  n0 <- as.integer(min(length(xy.out$x),length(xy.out$y)))
  x0 <- as.double(xy.out$x[1:n0])
  y0 <- as.double(xy.out$y[1:n0])
  z0 <- as.double(rep(0,n0))
  ixf <- iyf <- as.integer(rep(0,n0))
  fxf <- fyf <- as.double(rep(0,n0))
  z11f <- z12f <- z21f <- z22f <- as.double(rep(0,n0))

  nx <- as.integer(length(xyz.inp$x))
  ny <- as.integer(length(xyz.inp$y))

  xg <- as.double(xyz.inp$x)
  yg <- as.double(xyz.inp$y)
  zg <- as.double(as.vector( xyz.inp$z ))

  ##  zg[is.na(zg)] <- as.double(0)
  ## 4/25/08
  ##  Deal with the case when the path we are interpolating onto goes
  ##  outside the range of available data. See aterpfun.r
  nacon <- -1e30
  zg[is.na(zg)] <- as.double(nacon)

  nmax <- 800  ## see closest.f
  
  if ( length(zg) > nmax*nmax ) return('[Rneshaps::closest] Grid too large')
  if ( n0 > nmax ) return(paste('[Rneshaps::closest] Trying to interpolate too many points',n0,' (max =',nmax,')'))

##      subroutine closest(x0,y0,z0,n0,xg,yg,zg,nx,ny,ixf,iyf,fxf,fyf,
##     .     z11f,z12f,z21f,z22f,nacon)

  
  retn.full <- .Fortran("closest",
                        x0,y0,z0,n0,
                        xg,yg,zg,
                        nx,ny,
                        ixf,iyf,
                        fxf,fyf,
                        z11f,z12f,z21f,z22f,
                        nacon
                        )
  retn.full[[3]][retn.full[[3]] <= nacon/100] <- NA

  retn.many <- list(
                 x=retn.full[[1]],
                 y=retn.full[[2]],
                 z=retn.full[[3]],
                 ixf=retn.full[[10]],
                 iyf=retn.full[[11]],
                 fxf=retn.full[[12]],
                 fyf=retn.full[[13]],
                 z11f=retn.full[[14]],
                 z12f=retn.full[[15]],
                 z21f=retn.full[[16]],
                 z22f=retn.full[[17]],
                 nacon=nacon
                 )
  retn.few <- list(
                   x=retn.full[[1]],y=retn.full[[2]],z=retn.full[[3]]
		   )
  if (info) return(retn.many)
  retn.few
}

File: pkg/src/closest.f
      subroutine closest(x0,y0,z0,n0,xg,yg,zg,nx,ny,ixf,iyf,fxf,fyf,
     .     z11f,z12f,z21f,z22f,nacon)

c     LLNL-CODE-637312

c     subroutine to get z value at arbitrary x,y locations
c     from z values at closest location on regular rectangular grid, as returned by
c     the splus function interp()

c     x0,y0 locations at which interpolated values are desired
c     z0    place to put interpolated values
c     n0    number or x0,y0 pairs
c     xg,yg   x and y coordinates of grid to intepolate from
c     nx,ny   number of x and y values
c     zg      one-dimensional array of length nx*ny containing
c             z values for each x,y pair. zg is actually an
c             nx by ny matrix unrolled into a long 1-d array.

c     size arrays big enough for practical use
c     assume that the splus function calling this routine will test
c     assumes that x0 and y0 are sorted smallest to largest

c     4/25/04 supply a very large negative number (nacon) when input
c     grid matrix has NA in R. Then return the same (nacon) whenever
c     any adjacent grid location is nacon. This handles the case when
c     the path we are interpolating onto goes outside the range of available data

c     10/25/12 increase dimensions to 800, 640000

      double precision x0(800),y0(800),z0(800)
      double precision fxf(800),fyf(800)
      integer ixf(800),iyf(800)
      double precision z11f(800),z12f(800),z21f(800),z22f(800)
      double precision xg(800),yg(800),zg(640000)
      double precision nacon

      do 100 i0=1,n0
         x=x0(i0)
         y=y0(i0)
         do 200 ix=1,nx-1
            if (x .ge. xg(ix) .and. x .lt. xg(ix+1)) then
               do 300 iy=1,ny-1
                  if (y .ge. yg(iy) .and. y .lt. yg(iy+1)) then
                     fx=(x-xg(ix))/(xg(ix+1)-xg(ix))
                     fy=(y-yg(iy))/(yg(iy+1)-yg(iy))
                     z11=zg((iy-1)*nx+ix)
                     z12=zg(iy*nx+ix)
                     z21=zg((iy-1)*nx+ix+1)
                     z22=zg(iy*nx+ix+1)
                     if (z11 .le. nacon .or.
     x                    z12 .le. nacon .or.
     x                    z21 .le. nacon .or.
     x                    z22 .le. nacon) then
                        z0(i0)=nacon
                     else
                        if ( (fx .le. 0.5) .and. (fy .le. 0.5) ) then
                           z0(i0) = z11
                       elseif (( fx .gt. 0.5) .and. (fy .le. 0.5) ) then
                           z0(i0) = z12
                       elseif ( (fx .le. 0.5) .and. (fy .gt. 0.5) ) then
                           z0(i0) = z21 
                       elseif ( (fx .gt. 0.5) .and. (fy .gt. 0.5) ) then
                           z0(i0) = z22
                        endif
                     endif
                     ixf(i0)=ix
                     iyf(i0)=iy
                     fxf(i0)=fx
                     fyf(i0)=fy
                     z11f(i0)=z11
                     z12f(i0)=z12
                     z21f(i0)=z21
                     z22f(i0)=z22
                     goto 100
                     endif
 300           continue
               endif
 200      continue
 100   continue
       return
       end
